roaring64/bsi64.go

```diff
@@ -0,0 +1,498 @@
+// binary value.  The advantage of a BSI is that comparisons can be made
+// across ranges of values whereas a bitmap can only represent the existence
+// of a single value for a given column ID.  Another usage scenario involves
+// storage of high cardinality values.
+//
+// It depends upon the bitmap libraries.  It is not thread safe, so
+// upstream concurrency guards must be provided.
+type BSI struct {
+	bA       []*Bitmap
+	eBM      *Bitmap // Existence BitMap
+	MaxValue int64
+	MinValue int64
+}
+
+// NewBSI constructs a new BSI.  Min/Max values are optional.  If set to 0
+// then the underlying BSI will be automatically sized.
+func NewBSI(maxValue int64, minValue int64) *BSI {
+
+	ba := make([]*Bitmap, bits.Len64(uint64(maxValue)))
+	for i := 0; i < len(ba); i++ {
+		ba[i] = NewBitmap()
+	}
+	return &BSI{bA: ba, eBM: NewBitmap(), MaxValue: maxValue, MinValue: minValue}
+}
+
+// NewDefaultBSI constructs an auto-sized BSI
+func NewDefaultBSI() *BSI {
+	return NewBSI(int64(0), int64(0))
+}
+
+// ValueExists tests whether the value exists.
+func (b *BSI) ValueExists(columnID uint64) bool {
+
+	return b.eBM.Contains(uint64(columnID))
+}
+
+// GetCardinality returns a count of unique column IDs for which a value has been set.
+func (b *BSI) GetCardinality() uint64 {
+	return b.eBM.GetCardinality()
+}
+
+// BitCount returns the number of bits needed to represent values.
+func (b *BSI) BitCount() int {
+
+	return len(b.bA)
+}
+
+// SetValue sets a value for a given columnID.
+func (b *BSI) SetValue(columnID uint64, value int64) {
+
+	// If max/min values are set to zero then automatically determine bit array size
+	if b.MaxValue == 0 && b.MinValue == 0 {
+		ba := make([]*Bitmap, bits.Len64(uint64(value)))
+		for i := len(ba) - b.BitCount(); i > 0; i-- {
+			b.bA = append(b.bA, NewBitmap())
+		}
+	}
+
+	var wg sync.WaitGroup
+
+	for i := 0; i < b.BitCount(); i++ {
+		wg.Add(1)
+		go func(j int) {
+			defer wg.Done()
+			if uint64(value)&(1<<uint64(j)) > 0 {
+				b.bA[j].Add(uint64(columnID))
+			} else {
+				b.bA[j].Remove(uint64(columnID))
+			}
+		}(i)
+	}
+	wg.Wait()
+	b.eBM.Add(uint64(columnID))
+}
+
+// GetValue gets the value at the column ID.  Second param will be false for non-existant values.
+func (b *BSI) GetValue(columnID uint64) (int64, bool) {
+	value := int64(0)
+	exists := b.eBM.Contains(uint64(columnID))
+	if !exists {
+		return value, exists
+	}
+	for i := 0; i < b.BitCount(); i++ {
+		if b.bA[i].Contains(uint64(columnID)) {
+			value |= (1 << uint64(i))
+		}
+	}
+	return int64(value), exists
+}
+
+type action func(t *task, batch []uint64, resutsChan chan *Bitmap, wg *sync.WaitGroup)
+
+func parallelExecutor(parallelism int, t *task, e action,
+	foundSet *Bitmap) *Bitmap {
+
+	var n int = parallelism
+	if n == 0 {
+		n = runtime.NumCPU()
+	}
+
+	resultsChan := make(chan *Bitmap, n)
+
+	card := foundSet.GetCardinality()
+	x := uint64(math.Floor(float64(card) / float64(n)))
```

Comment:
I know that you have used this construction before, but this seems unnecessarily complicated. Given that card and n are non-negative, then you can just do x := card / n, minus some casting. The integer division is already defined as a floor.